### PR TITLE
Formalized repository URL parsing and resolving.

### PR DESCRIPTION
**Pull request by isoos**
- reusing and repurposing the tests for the now-defunct `getRepositoryUrl`

### CODE REVIEW

File: lib/src/repository_url.dart [hunk 1]
```diff
@@ -0,0 +1,180 @@
...
+  static const github = 'github';
+  static const gitlab = 'gitlab';
+  static const unknown = 'unknown';
+
+  static bool isGitHubCompatible(String provider) =>
+      provider == github || provider == gitlab;
+}
+
+const _repoReplacePrefixes = {
+  'http://github.com': 'https://github.com',
```
Comment (sigurdm) [inline lib/src/repository_url.dart:98]: Perhaps this replacement is better done by changing the scheme of the parsed url.

```
url = url.replace(scheme:'https')
``

File: lib/src/repository_url.dart [hunk 2]
```diff
@@ -0,0 +1,180 @@
...
+      .pathSegments
+      .where((s) => s.isNotEmpty)
+      .toList();
+  final repoSegmentIndex = _repoSegmentIndex(provider, segments);
+  if (repoSegmentIndex < 0) return null;
+
+  String? separator;
+  final extraSegments = segments.skip(repoSegmentIndex + 1).toList();
+  if (extraSegments.isNotEmpty &&
+      _githubSegmentSeparators.contains(extraSegments.first)) {
```
Comment (sigurdm) [inline lib/src/repository_url.dart:131]: Should this also be conditioned on the provider being github?

**Review by sigurdm — APPROVED** [review]
LGTM

File: lib/src/repository_url.dart (hunk 1 above)
Comment (isoos) [reply lib/src/repository_url.dart]: Done.

File: lib/src/repository_url.dart (hunk 2 above)
Comment (isoos) [reply lib/src/repository_url.dart]: Done.

**Review by sigurdm — APPROVED** [review]
LGTM

**Pull request merged**